packages/grid/x-data-grid-pro/src/hooks/features/lazyLoader/useGridLazyLoader.ts

```diff
@@ -0,0 +1,237 @@
+    return true;
+  }
+
+  if (rowsLoadingMode !== GridFeatureModeConstant.server) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * @requires useGridRows (state)
+ * @requires useGridPagination (state)
+ * @requires useGridDimensions (method) - can be after
+ * @requires useGridScroll (method
+ */
+export const useGridLazyLoader = (
+  apiRef: React.MutableRefObject<GridApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    | 'onFetchRows'
+    | 'rowsLoadingMode'
+    | 'pagination'
+    | 'paginationMode'
+    | 'rowBuffer'
+    | 'experimentalFeatures'
+  >,
+): void => {
+  const visibleRows = useGridVisibleRows(apiRef, props);
+  const rowIds = useGridSelector(apiRef, gridVisibleSortedRowIdsSelector);
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const filterModel = useGridSelector(apiRef, gridFilterModelSelector);
+  const renderedRowsIntervalCache = React.useRef<GridRenderedRowsIntervalChangeParams>({
+    firstRowToRender: 0,
+    lastRowToRender: 0,
+  });
+  const { lazyLoading } = (props.experimentalFeatures ?? {}) as GridExperimentalProFeatures;
+
+  const getCurrentIntervalToRender = React.useCallback((): {
+    firstRowToRender: number;
+    lastRowToRender: number;
+  } => {
+    const currentRenderContext = apiRef.current.unstable_getRenderContext();
+    const [firstRowToRender, lastRowToRender] = getRenderableIndexes({
+      firstIndex: currentRenderContext.firstRowIndex,
+      lastIndex: currentRenderContext.lastRowIndex,
+      minFirstIndex: 0,
+      maxLastIndex: visibleRows.rows.length,
+      buffer: props.rowBuffer,
+    });
+
+    return {
+      firstRowToRender,
+      lastRowToRender,
+    };
+  }, [apiRef, props.rowBuffer, visibleRows.rows.length]);
+
+  const handleRenderedRowsIntervalChange = React.useCallback(
+    (params: GridRenderedRowsIntervalChangeParams) => {
+      const dimensions = apiRef.current.getRootDimensions();
+
+      if (
+        isLazyLoadingDisabled({
+          lazyLoadingFeatureFlag: lazyLoading,
+          rowsLoadingMode: props.rowsLoadingMode,
+          gridDimentions: dimensions,
+        })
+      ) {
+        return;
+      }
+
+      if (
+        renderedRowsIntervalCache.current.firstRowToRender === params.firstRowToRender &&
+        renderedRowsIntervalCache.current.lastRowToRender === params.lastRowToRender
+      ) {
+        return;
+      }
+
+      const renderedRowsIds: Array<GridRowId | null> = [...rowIds].splice(
+        params.firstRowToRender,
+        params.lastRowToRender - params.firstRowToRender,
+      );
+      const hasSkeletonRowIds = renderedRowsIds.some((rowId) =>
+        `${rowId}`.includes(GRID_SKELETON_ROW_ROOT_ID),
+      );
+
+      if (!hasSkeletonRowIds) {
+        return;
+      }
+
+      const { firstRowIndex, lastRowIndex } = findSkeletonRowsSection(visibleRows.rows, {
+        firstRowIndex: params.firstRowToRender,
+        lastRowIndex: params.lastRowToRender,
+      });
```

Comment:
It can become complicated. The `hasSkeletonRowIds` pourpos is to know if there is a skeleton row in the current rendered and sorted `rowId`'s.

Reply:
Yes, but `findSkeletonRowsSection` would also tell whether there are skeleton rows or not.
It operates in the same bounds of `[findSkeletonRowsSection, lastRowToRender]` as `hasSkeletonRowIds`.

Consider the use case I've described above - it would be much faster to use `findSkeletonRowsSection` without even trying to evaluate `hasSkeletonRowIds`.